src/xc_integrator/xc_data/device/xc_device_data.hpp

```diff
@@ -60,25 +76,40 @@ struct required_term_storage {
   bool grid_vrho     = false;
   bool grid_vgamma   = false;
 
+
+  // Reference flags for memory management use
+  integrator_term_tracker ref_tracker;
+
   inline size_t grid_den_size(size_t npts){ 
-    return PRDVL(grid_den, npts);
+  // grid_den_size takes into account the size of the interleaved density sent to ExchCXX in the cases of UKS/GKS (hence the * 2)
+    if( grid_den ) {
+      if( ref_tracker.ks_scheme == UKS) return 2 * npts;
+      else return npts;
+    }
+    else return 0ul;
   }
   inline size_t grid_den_grad_size(size_t npts){ 
-    return PRDVL(grid_den_grad, 3 * npts);
+    return PRDVL(grid_den_grad and ref_tracker.ks_scheme == RKS, 3 * npts)
+         + PRDVL(grid_den_grad and ref_tracker.ks_scheme == UKS, 6 * npts);
   }
   inline size_t grid_gamma_size(size_t npts){ 
-    return PRDVL(grid_gamma, npts);
+    return PRDVL(grid_gamma and ref_tracker.ks_scheme == RKS, npts) 
+         + PRDVL(grid_gamma and ref_tracker.ks_scheme == UKS, 3 * npts);
   }
   inline size_t grid_eps_size(size_t npts){ 
     return PRDVL(grid_eps, npts);
   }
   inline size_t grid_vrho_size(size_t npts){ 
-    return PRDVL(grid_vrho, npts);
+    return PRDVL(grid_vrho and ref_tracker.ks_scheme == RKS, npts)
+         + PRDVL(grid_vrho and ref_tracker.ks_scheme == UKS, 2 * npts);
   }
   inline size_t grid_vgamma_size(size_t npts){ 
-    return PRDVL(grid_vgamma, npts);
+    return PRDVL(grid_vgamma and ref_tracker.ks_scheme == RKS, npts)
+         + PRDVL(grid_vgamma and ref_tracker.ks_scheme == UKS, 3 * npts);
```

Comment:
These should just be populated as booleans at construction and used whenever they're needed (e.g. `is_rks`, `is_uks`, etc)